### PR TITLE
[ISSUE-611][BUG] FetchHandler should handle PartitionFileSorter return null and we should enable retry for sorter exception

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### Related issues.
https://github.com/alibaba/RemoteShuffleService/issues/611

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java [hunk 1]
```diff
@@ -217,7 +217,10 @@ private synchronized boolean shouldRetry(Throwable e) {
             || e instanceof TimeoutException
             || (e.getCause() != null && e.getCause() instanceof TimeoutException)
             || (e.getCause() != null && e.getCause() instanceof IOException);
-    return isIOException && hasRemainingRetries();
+    boolean isSortException =
+        e instanceof RuntimeException
+            && e.getMessage().startsWith("com.aliyun.emr.rss.service.deploy.worker.exception");
+    return (isIOException || isSortException) && hasRemainingRetries();
```
Comment (AngersZhuuuu) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java:223]: I am not sure if it's ok to do like this, but at least, here miss to handle `openStream` return `null`.

And we can avoid return null and show clear error message for each case

**Review by FMX — APPROVED** [review]
LGTM. Thanks.

File: client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java (hunk 1 above)
Comment (FMX) [reply client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java:223]: I think this section has some problems. SortedAccessHDFSException is not a subclass of RuntimeException. So this section cannot work properly.

**Review by FMX — CHANGES_REQUESTED** [review]
update status.

File: client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java [hunk 2]
```diff
@@ -217,7 +217,10 @@ private synchronized boolean shouldRetry(Throwable e) {
             || e instanceof TimeoutException
             || (e.getCause() != null && e.getCause() instanceof TimeoutException)
             || (e.getCause() != null && e.getCause() instanceof IOException);
-    return isIOException && hasRemainingRetries();
+    boolean isSortException =
```
Comment (waitinfuture) [inline client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java:220]: We can let all new added Exceptions inherit from IOException, so we needn't change here.

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/storage/PartitionFilesSorter.java [hunk 3]
```diff
@@ -158,7 +159,9 @@ public int getSortedSize() {
   }
 
   public FileInfo openStream(
-      String shuffleKey, String fileName, FileInfo fileInfo, int startMapIndex, int endMapIndex) {
+      String shuffleKey, String fileName, FileInfo fileInfo, int startMapIndex, int endMapIndex)
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/storage/PartitionFilesSorter.java:162]: throws IOException

File: client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java (hunk 2 above)
Comment (AngersZhuuuu) [reply client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java:220]: > We can let all new added Exceptions inherit from IOException, so we needn't change here.

Nice suggestion!

File: client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java (hunk 1 above)
Comment (AngersZhuuuu) [reply client/src/main/java/com/aliyun/emr/rss/client/read/RetryingChunkClient.java]: > I think this section has some problems. SortedAccessHDFSException is not a subclass of RuntimeException. So this section cannot work properly.

When exception pass to Client side, it had been wrapped as a RPCFailure
```
new RpcFailure(
          request.requestId,
          Throwables.getStackTraceAsString(fnf))
```

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/storage/PartitionFilesSorter.java (hunk 3 above)
Comment (AngersZhuuuu) [reply server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/storage/PartitionFilesSorter.java:161]: Yea, done

**Comment by AngersZhuuuu** [comment]
How about current? @waitinfuture @FMX

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Review by FMX — APPROVED** [review]

**Pull request merged**